File: 120-triangle/triangle.java
class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        if(triangle == null || triangle.isEmpty()) return 0;

        for(int i = triangle.size() - 2; i >= 0; i--){
            List<Integer> currentRow = triangle.get(i);
            List<Integer> nextRow = triangle.get(i+1);

            for(int j = 0; j < currentRow.size(); j++){
                int sum1 = nextRow.get(j);
                int sum2 = nextRow.get(j + 1);

                int newSum = currentRow.get(j) + Math.min(sum1, sum2);
                currentRow.set(j, newSum);
            }
        }
        return triangle.get(0).get(0);
    }
}

File: 120-triangle/triangle.py
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        def function(m, n, dp, k, triangle):
            if dp[m][n] != None:
                return dp[m][n]

            if m == k - 1:
                return triangle[m][n]

            left = triangle[m][n] + function(m + 1, n, dp, k, triangle)
            right = triangle[m][n] + function(m + 1, n + 1, dp, k, triangle)

            dp[m][n] = min(left, right)
            return dp[m][n]

        rows = len(triangle)
        dp = [[None] * rows for _ in range(rows)]

        return function(0, 0, dp, rows, triangle)